msprime/simulations.py

```diff
@@ -1365,6 +1366,167 @@ def print_history(self, output=sys.stdout):
+            This should be set to a value smaller than the smallest
+            desired population size in the model.
+
+        :return: Returns a tuple where the first value is an array of time points,
+            the second is an array whose jth element is the coalescence
+            rate at the jth time point (denoted r(t[j]) above), and the third is
+            the probablility that a randomly pair of lineages has not yet coalesced
+            (denoted p(t[j]) above).
+
+        :rtype: class:'tuple'
+        """
+        num_pops = self.num_populations
+        # TODO throw a useful error if this fails.
+        assert(len(num_samples) == num_pops)
+        P = np.zeros([num_pops**2, num_pops**2])
+        IA = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        Identity = np.eye(num_pops)
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples[x]
+                K_y = num_samples[y]
+                P[IA[x, y], IA[x, y]] = K_x * (K_y - (x == y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        p_t = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        mass_migration_objects = []
+        mass_migration_steps = []
+        for demo in self.demographic_events:
+            if type(demo) == MassMigration:
+                mass_migration_objects.append(demo)
+                mass_migration_steps.append(int(demo.time/dt))
+        for j in range(num_steps):
+            N_M = self.pop_size_and_migration_at_t(steps[j])
+            N = N_M["population_sizes"]
+            M = N_M["migration_matrix"]
+            C = np.zeros([num_pops**2, num_pops**2])
+            for idx in range(num_pops):
+                C[IA[idx, idx], IA[idx, idx]] = 1 / (2 * max(min_pop_size, N[idx]))
+            dM = np.diag([sum(s) for s in M])
+            G = (np.kron(M - dM, Identity) + np.kron(Identity, M - dM)) - C
+            if j in mass_migration_steps:
+                idx = mass_migration_steps.index(j)
+                ds = mass_migration_objects[idx].time - (j * dt)
+                a = mass_migration_objects[idx].source
+                b = mass_migration_objects[idx].dest
+                p = mass_migration_objects[idx].proportion
+                S = np.eye(num_pops**2, num_pops**2)
+                for x in range(num_pops):
+                    if x == a:
+                        S[IA[a, a], IA[a, b]] = S[IA[a, a], IA[b, a]] = p * (1 - p)
+                        S[IA[a, a], IA[b, b]] = p ** 2
+                        S[IA[a, a], IA[a, a]] = (1 - p) ** 2
+                    else:
+                        S[IA[x, a], IA[x, b]] = S[IA[a, x], IA[b, x]] = p
+                        S[IA[x, a], IA[x, a]] = S[IA[a, x], IA[a, x]] = 1 - p
+                P = np.matmul(P, self._expm(ds * G))
+                P = np.matmul(P, S)
+                P = np.matmul(P, self._expm((dt - ds) * G))
```

Comment:
This is a bug: we're still using the old `G` for the first (dt - ds) of the new epoch. In the Gutenkunst et al model, at time 848, there's a mass migration from 2 -> 1 *and* concurrently all the migration rates to 2 get set to zero, so no lineages can go back there. But, since we don't update `G`, some lineages sneak back into 2.